tools/sof-perf-analyzer.py

```diff
@@ -0,0 +1,253 @@
+#!/usr/bin/env python3
+
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright(c) 2023 Intel Corporation. All rights reserved.
+
+'''
+The sof_trace_analyzer helps to analyze SOF trace output generated during the running of
+aplay and arecord commands or a test case in sof-test repository. It takes human readable
+trace file (contents from plain mtrace or the decoded dictionary trace) as input, and
+output below analysis results:
+    - Average and maximum MCPS of cpu average and cpu peak per SOF component
+
+Hardcoded value explained:
+DSP_CLK: The DSP clock frequency in Hz.
+DSP_TIMER: The DSP timer clock frequency in Hz.
+'''
+
+import re
+import pathlib
+import argparse
+from typing import Dict
+from typing import List
+from typing import TextIO
+from typing import Optional
+from typing import Generator
+from dataclasses import dataclass
+
+@dataclass()
+class TraceItem:
+    '''The structural representation for a single line of trace'''
+    timestamp: float
+    level: str
+    context: str # the trace context registered with LOG_MODULE_REGISTER
+    func: str
+    msg: str
+
+@dataclass(frozen=True)
+class Comp:
+    '''The identifier for a SOF component'''
+    ppl_id: str
+    comp_id: str
+
+@dataclass(frozen=True)
+class CompPerfInfo:
+    '''The dataclass for SOF component performance info'''
+    timestamp: float
+    samples: int
+    period: int
+    cpu_avg: int
+    cpu_peak: int
+
+@dataclass(frozen=True)
+class CompPerfStats:
+    '''The dataclass for SOF component statistics'''
+    avg_cpu_avg: float   # average of cpu average
+    avg_cpu_peak: float  # average of cpu peak
+    peak_cpu_avg: float  # peak of cpu average
+    peak_cpu_peak: float # peak of cpu peak
+
+# Currently, we don't have test case to test low-power mode, in which
+# the DSP may not operate in the maximum clock frequency. Let't hardcode
+# this value to the maximum clock frequency for now.
+DSP_CLK = 400000000
+# The DSP timer clock frequency is fixed despite of DSP_CLK
+DSP_TIMER = 38400000
+UINT32_MAX = 4294967295
+AUDIO_PERIOD = 1000
+
+TICK_TO_MCPS = DSP_CLK / DSP_TIMER / AUDIO_PERIOD
+
+# Once DSP timer count to UINT32_MAX, it wrapped to zero, which causes
+# the timestamp wrap to zero, too. This ts_shift is used to correct
+# the timestamp value on wrap. UINT32_MAX / DSP_TIMER is added to it at
+# every wrap.
+#
+# pylint: disable=C0103
+ts_shift = 0
+# pylint: disable=C0103
+args = None
+
+# The dictionary for performance information
+perf_info: Dict[Comp, List[CompPerfInfo]] = {}
+# The dictionary for performace statistics
+perf_stats: Dict[Comp, CompPerfStats] = {}
+# Component ID to name map acquired from dmesg
+comp_name_map: Dict[Comp, str] = {}
```

Comment:
Most of the text in the comments merely duplicates the names.

```python
# This map gives access to the list of samples collected for a given component
perf_info: Dict[Comp, List[CompPerfInfo]] = {}
```

A `_map` suffix is almost never useful because it's almost always obvious from the definition and usage.

```python
components_name: Dict[...

...

components_name[my_component] ...
```